core/io/resource_loader.cpp

```diff
@@ -454,6 +454,23 @@ Ref<Resource> ResourceLoader::load(const String &p_path, const String &p_type_hi
 	return res;
 }
 
+// Returns a shallow reference of the loading resource.
+// Will fail if the resource is not loading.
+Ref<Resource> ResourceLoader::get_loading_resource(const String &p_path, Error *r_error) {
+	ERR_FAIL_COND_V(r_error == nullptr, nullptr);
```

Comment:
If passing is required I'd pass a reference instead to make the requirement explicit

Reply:
While we're at it, I'm not very sure about the "shallow" wording. Since this is async, by the time the caller deals with the resource, the completeness in its properties having been set and references to dependencies being filled ranges from none (we could call that "shallow") to full. Also, this wouldn't be thread safe, in the sense that there's no sync point with the potentially arbitrary loading thread, which can lead to the caller not only seeing an incomplete resource, but running into data races or race conditions.

Do you think it would be an option to wait for the full resource to be loaded? That would guarantee safety.

If the caller can be happy with the resource in a shallow state (that is, own properties set) the loader needs to be enhanced to support it properly. However, bear in mind that the caller can't safely access anything that can still change (references to external resources, properties already set but that will change due to one of those "queued update" callbacks running later, etc.).

Reply:
For GDScript this doesn't matter in most cases. Only if the user `preload()` a currently loading resource and do something with the resource at the point. This is probably only possible in a static initializer (explicit or implicit).

If we choose this approach, then we should add a comment that `ResourceLoader::get_loading_resource()` is intended for GDScript (due to how it works) and should not be used elsewhere in the engine.

Alternatively, we could make such a "shallow" resource `null` until the end of loading. In both cases it will look unexpected to the user. Not sure which option is better, but probably `null` option is more explicit.

> If the required non-GDScript resource is currently being loaded, then we should not call `ResourceLoader::load()`. But we need get it _sometime_, and its loading will complete later than the script. What could be bad if the script executes a code before that, for instance the static initializer.
>
> Here we have two options:
>
> 1. Add a hook to `ResourceLoader` that allow us to get the resource when it is fully loaded. Until than it is `null` in the script.
> 2. Add a posibility to `ResourceLoader` to get a shallow resource, i.e. a currently loading resource, that is not fully initialized yet.

Reply:
Adding a thought here: I think this only needs to be aware of other loads happening on the same thread (in other words, in the same stack). In that case, `ResourceLoader` could have a different function for that case explicitly that wouldn't involve any interaction with other threads. I could implement it if you think that'd do the trick.

Reply:
> Given there's no nuance to report about the error, I think returning a null `Ref` is enough, so `r_error` can go away.

It's impossible. 99% of the time, the returned reference **will** be null. Hence why I force the error return value.

Reply:
> If we choose this approach, then we should add a comment that `ResourceLoader::get_loading_resource()` is intended for GDScript (due to how it works) and should not be used elsewhere in the engine.

I don't think it's a good idea. Like, I think we could push this technique to support cyclic loading elsewhere too. But I agree that this function should have a warning or something like that.

Reply:
> Adding a thought here: I think this only needs to be aware of other loads happening on the same thread (in other words, in the same stack). In that case, `ResourceLoader` could have a different function for that case explicitly that wouldn't involve any interaction with other threads. I could implement it if you think that'd do the trick.

Elaborating on that:

Since this PR is about a script that is cyclically referenced, `ResourceLoader` already has a data member that fits the bill: `load_paths_stack`. That's unique per thread, so checking there is good as it keeps the concern within the dependency chain of the current load chain, deterministically as no threads are involved. However, that vector would need to be modified to it contains the resource references as well.

Reply:
> Since this PR is about a script that is cyclically referenced

This PR is more about anything that is cyclically referenced _in the context_ of script loading.